File: code/packages/python/nib-wasm-compiler/tests/test_nib_wasm_compiler.py
from __future__ import annotations

import pytest
from wasm_runtime import WasmRuntime

from nib_wasm_compiler import NibWasmCompiler, PackageError, compile_source, pack_source, write_wasm_file


def test_compile_source_returns_pipeline_artifacts() -> None:
    source = """
fn answer() -> u4 {
    return 7;
}
"""
    result = compile_source(source)
    assert result.raw_ir is not None
    assert result.optimized_ir is not None
    assert ".func" in result.wasm_assembly
    assert result.binary


def test_pack_source_is_alias_for_compile_source() -> None:
    source = "fn answer() -> u4 { return 7; }"
    compiled = compile_source(source)
    packed = pack_source(source)
    assert packed.binary == compiled.binary
    assert packed.wasm_assembly == compiled.wasm_assembly


def test_write_wasm_file_writes_output(tmp_path) -> None:
    source = "fn answer() -> u4 { return 7; }"
    output = tmp_path / "program.wasm"
    result = write_wasm_file(source, output)
    assert output.exists()
    assert output.read_bytes() == result.binary


def test_compiled_program_runs_in_wasm_runtime() -> None:
    source = """
fn add(a: u4, b: u4) -> u4 {
    return a +% b;
}

fn main() -> u4 {
    return add(3, 4);
}
"""
    result = compile_source(source)
    runtime_result = WasmRuntime().load_and_run(result.binary, "_start", [])
    assert runtime_result == [7]


def test_compiled_loop_runs_in_wasm_runtime() -> None:
    source = """
fn count_to(n: u4) -> u4 {
    let acc: u4 = 0;
    for i: u4 in 0..n {
        acc = acc +% 1;
    }
    return acc;
}
"""
    result = compile_source(source)
    runtime_result = WasmRuntime().load_and_run(result.binary, "count_to", [5])
    assert runtime_result == [5]


def test_compiled_loop_and_call_runs_in_wasm_runtime() -> None:
    source = """
fn add(a: u4, b: u4) -> u4 {
    return a +% b;
}

fn main() -> u4 {
    let acc: u4 = 0;
    for i: u4 in 0..3 {
        acc = add(acc, 1);
    }
    return acc;
}
"""
    result = compile_source(source)
    runtime_result = WasmRuntime().load_and_run(result.binary, "_start", [])
    assert runtime_result == [3]


def test_type_error_raises_package_error() -> None:
    source = "fn main() { let x: bool = 1 +% 2; }"
    with pytest.raises(PackageError) as exc_info:
        compile_source(source)
    assert exc_info.value.stage == "type-check"


def test_parse_error_raises_package_error() -> None:
    source = "fn main("
    with pytest.raises(PackageError) as exc_info:
        NibWasmCompiler().compile_source(source)
    assert exc_info.value.stage == "parse"

File: code/packages/ruby/nib_wasm_compiler/test/test_nib_wasm_compiler.rb
# frozen_string_literal: true

require_relative "test_helper"

class NibWasmCompilerTest < Minitest::Test
  def runtime
    CodingAdventures::WasmRuntime::Runtime.new
  end

  def test_compile_source_returns_binary
    result = CodingAdventures::NibWasmCompiler.compile_source("fn answer() -> u4 { return 7; }")

    assert result.binary.bytesize.positive?
    assert result.raw_ir.instructions.any?
  end

  def test_pack_source_aliases_compile_source
    compiled = CodingAdventures::NibWasmCompiler.compile_source("fn answer() -> u4 { return 7; }")
    packed = CodingAdventures::NibWasmCompiler.pack_source("fn answer() -> u4 { return 7; }")

    assert_equal compiled.binary, packed.binary
  end

  def test_write_wasm_file_persists_bytes
    Dir.mktmpdir do |dir|
      path = File.join(dir, "program.wasm")
      result = CodingAdventures::NibWasmCompiler.write_wasm_file("fn answer() -> u4 { return 7; }", path)

      assert_equal result.binary, File.binread(path)
      assert_equal path, result.wasm_path
    end
  end

  def test_compiled_program_runs_through_start
    result = CodingAdventures::NibWasmCompiler.compile_source(<<~NIB)
      fn add(a: u4, b: u4) -> u4 { return a +% b; }
      fn main() -> u4 { return add(3, 4); }
    NIB

    assert_equal [7], runtime.load_and_run(result.binary, "_start", [])
  end

  def test_compiled_loop_runs_through_export
    result = CodingAdventures::NibWasmCompiler.compile_source(<<~NIB)
      fn count_to(n: u4) -> u4 {
        let acc: u4 = 0;
        for i: u4 in 0..n {
          acc = acc +% 1;
        }
        return acc;
      }
    NIB

    assert_equal [5], runtime.load_and_run(result.binary, "count_to", [5])
  end

  def test_type_errors_raise_package_error
    error = assert_raises(CodingAdventures::NibWasmCompiler::PackageError) do
      CodingAdventures::NibWasmCompiler.compile_source("fn main() { let x: bool = 1 +% 2; }")
    end

    assert_equal "type-check", error.stage
  end
end
